File: Top_Category_Codes/SoftwareLo1988354192006/SoftwareLock/CS/SoftwareLock.cs
using System.Diagnostics;
using System.Data;
using System.Collections;
using Microsoft.VisualBasic;
using System.Collections.Generic;
using System;
using System.Security.Cryptography;
using Microsoft.Win32;
using System.Text;

//I'll be using Microsoft's Cryptography Class
//for encryption purposes:

//All registration data is stored to windows
//Registry:

//I need UnicodeEncoding class of
//this namespace:

namespace SoftwareLock
{
	public partial class SoftwareLock
	{
		
		
		protected string _AppName;
		protected string _Password;
		protected HashAlgorithms _HashAlgorithm;
		protected string _DiskSerial;
		protected bool _isRegistered;
		protected RegistryHives _RegistryHive;
		protected string _RegistryBaseKey;
		protected string _RegistrySubKey;
		protected string _CustRef;
		protected string _SerialKey;
		protected int _CustRefLength;
		public RegistryHives RegistryHive
		{
			get
			{
				return _RegistryHive;
			}
			set
			{
				_RegistryHive = value;
			}
		}
		
		public enum HashAlgorithms
		{
			MD5 = 0,
			SHA1 = 1,
			SHA256 = 2,
			SHA384 = 3,
			SHA512 = 4
		}
		
		public enum RegistryHives
		{
			CURRENTUSER,
			LOCALMACHINE
		}
		
		public SoftwareLock()
		{
			_DiskSerial = HDDSerial.SerialNumber(HDDSerial.GetCurrentDrive());
			_CustRefLength = 20;
		}
		
		public string AppName
		{
			get
			{
				return _AppName;
			}
			set
			{
				_AppName = value;
			}
		}
		
		public string Password
		{
			get
			{
				return _Password;
			}
			set
			{
				_Password = value;
			}
		}
		
		//return of set HashAlgorithm for generating
		//Reference and serialkey
		public HashAlgorithms HashAlgorithm
		{
			get
			{
				return _HashAlgorithm;
			}
			set
			{
				_HashAlgorithm = value;
			}
		}
		
		//returns end-user's hard disk serial number
		public string HardDiskSerial
		{
			get
			{
				return _DiskSerial;
			}
		}
		
		//use this property to check if your software is registered
		public bool isRegistered
		{
			
			get
			{
				
				if (_RegistryBaseKey.Trim() == "")
				{
					throw (new Exception("Specify a BaseRegistryKey"));
				}
				
				try
				{
					Microsoft.Win32.RegistryKey objRegistry;
					switch (_RegistryHive)
					{
						case RegistryHives.CURRENTUSER:
							
							objRegistry = Registry.CurrentUser;
							break;
						case RegistryHives.LOCALMACHINE:
							
							objRegistry = Registry.LocalMachine;
							break;
						default:
							
							objRegistry = Registry.CurrentUser;
							break;
					}
					
					RegistryKey RK = objRegistry. OpenSubKey("Software\\" + _RegistryBaseKey);
					string SKey = RK.GetValue(_RegistrySubKey);
					if (SKey == SerialKey)
					{
						return true;
					}
					else
					{
						return false;
					}
				}
				catch (NullReferenceException)
				{
					return false;
				}
				
			}
			
		}
		
		//Base Registry key:
		//usually your software name
		public string RegistryBaseKey
		{
			get
			{
				return _RegistryBaseKey;
			}
			set
			{
				_RegistryBaseKey = value;
			}
		}
		
		//a sub key of BaseRegistryKey to store
		//the software's serial key
		public string RegistrySubKey
		{
			get
			{
				return _RegistrySubKey;
			}
			set
			{
				_RegistrySubKey = value;
			}
		}
		
		public int ReferenceLength
		{
			get
			{
				return _CustRefLength;
			}
			set
			{
				if (value >= 6 && value <= 20)
				{
					_CustRefLength = value;
				}
				else
				{
					throw (new Exception("ReferenceLength cannot be less than 6 or more than 20 characters"));
				}
			}
		}
		
		//use this method in your software's registration form's
		//OK button, and pass it the serial key entered by the end-user
		//as argument, returns True for a successful registration, False
		//for unsuccessful
		public bool Register(string strSerialKey)
		{
			GenerateCodes();
			
			Microsoft.Win32.RegistryKey objRegistry;
			switch (_RegistryHive)
			{
				
			case RegistryHives.CURRENTUSER:
				
				objRegistry = Registry.CurrentUser;
				break;
			case RegistryHives.LOCALMACHINE:
				
				objRegistry = Registry.LocalMachine;
				break;
			default:
				
				objRegistry = Registry.CurrentUser;
				break;
		}
		
		if (strSerialKey == _SerialKey)
		{
			RegistryKey RK = objRegistry. OpenSubKey("Software", true);
			RK.CreateSubKey(_RegistryBaseKey). SetValue(_RegistrySubKey, _SerialKey);
			return true;
		}
		else
		{
			return false;
		}
	}
	
	//returns customer reference based on appname,password,harddisk serial
	virtual public string Reference
	{
		get
		{
			GenerateCodes();
			return _CustRef;
		}
		set
		{
			if (value.Length >= 6 && value.Length <= 20)
			{
				_CustRef = value;
			}
			else
			{
				throw (new Exception("Reference cannot be less than 6 or more than 20 characters"));
			}
		}
	}
	
	//returns serial key based on Reference
	virtual public string SerialKey
	{
		get
		{
			GenerateCodes();
			return _SerialKey;
		}
	}
	
	public virtual bool CancelRegistration()
	{
		//you can use this function for testing purposes
		//should you need to remove any test entries from
		//the windows registry where registration information is stored
		
		RegistryKey objRegistry;
		switch (_RegistryHive)
		{
			case RegistryHives.CURRENTUSER:
				
				objRegistry = Registry.CurrentUser;
				break;
			case RegistryHives.LOCALMACHINE:
				
				objRegistry = Registry.LocalMachine;
				break;
			default:
				
				objRegistry = Registry.CurrentUser;
				break;
		}
		
		RegistryKey RK = objRegistry.OpenSubKey("Software", true);
		Interaction.MsgBox("The followinag key and all its subkeys will be deleted:" + "\r\n" + RK.ToString() + "\\" + _RegistryBaseKey, MsgBoxStyle.Information, null);
		RK.DeleteSubKeyTree(_RegistryBaseKey);
		
	}
	
	protected void GenerateCodes()
	{
		//generate Reference based on
		//appname,password,and harddisk serial
		//generate serial key based on Reference
		HashAlgorithm Hash;
		byte[] HashBytes;
		UnicodeEncoding UNIEncoding = new UnicodeEncoding();
		string temp;
		
		//load the user selected hash-algorithm
		switch (_HashAlgorithm)
		{
			case HashAlgorithms.MD5:
				
				Hash = new MD5CryptoServiceProvider();
				break;
			case HashAlgorithms.SHA1:
				
				Hash = new SHA1CryptoServiceProvider();
				break;
			case HashAlgorithms.SHA256:
				
				Hash = new SHA256Managed();
				break;
			case HashAlgorithms.SHA384:
				
				Hash = new SHA384Managed();
				break;
			case HashAlgorithms.SHA512:
				
				Hash = new SHA512Managed();
				break;
			default: //default hash algorithm
				
				Hash = new MD5CryptoServiceProvider();
				break;
		}
		
		//generate hash using appname, password, and HD serial no.
		temp = _AppName + _Password + _DiskSerial;
		HashBytes = Hash.ComputeHash(UNIEncoding.GetBytes(temp));
		
		_CustRef = Convert.ToBase64String(HashBytes);
		//if longer than 20 chars, trim it to 20 chars
		if (_CustRef.Trim().Length > _CustRefLength)
		{
			_CustRef = _CustRef.Substring(0, _CustRefLength);
		}
		
		//convert it to upper case
		_CustRef = _CustRef.ToUpper();
		
		//clear the hash array
		Array.Clear(HashBytes, 0, HashBytes.Length);
		
		//use the above generated Reference to generate a 20 characters
		//serial key for the end user
		temp = _CustRef + _AppName + _Password;
		HashBytes = Hash.ComputeHash(UNIEncoding.GetBytes(temp));
		
		_SerialKey = Convert.ToBase64String(HashBytes);
		if (_SerialKey.Trim().Length > 20)
		{
			_SerialKey = _SerialKey.Substring(0, 20);
		}
		_SerialKey = _SerialKey.ToUpper();
		
	}
	
	
}

}

File: Top_Category_Codes/SoftwareLo1988354192006/SoftwareLock/SoftwareLock.vb
'I'll be using Microsoft's Cryptography Class
'for encryption purposes:
Imports System.Security.Cryptography

'All registration data is stored to windows
'Registry:
Imports Microsoft.Win32

'I need UnicodeEncoding class of
'this namespace:
Imports System.Text

  Public Class SoftwareLock

  Protected _AppName As String
  Protected _Password As String
  Protected _HashAlgorithm As HashAlgorithms
  Protected _DiskSerial As String
  Protected _isRegistered As Boolean
  Protected _RegistryHive As RegistryHives
  Protected _RegistryBaseKey As String
  Protected _RegistrySubKey As String
  Protected _CustRef As String
  Protected _SerialKey As String
  Protected _CustRefLength As Integer
  Public Property RegistryHive() As RegistryHives
    Get
      Return _RegistryHive
    End Get
    Set(ByVal value As RegistryHives)
      _RegistryHive = value
    End Set
  End Property

  Public Enum HashAlgorithms
    MD5 = 0
    SHA1 = 1
    SHA256 = 2
    SHA384 = 3
    SHA512 = 4
  End Enum

  Public Enum RegistryHives
    CURRENTUSER
    LOCALMACHINE
  End Enum

  Public Sub New()
    MyBase.New()
    _DiskSerial = SerialNumber(GetCurrentDrive())
    _CustRefLength = 20
  End Sub

  Property AppName() As String
    Get
      Return _AppName
    End Get
    Set(ByVal value As String)
      _AppName = value
    End Set
  End Property

  Property Password() As String
    Get
      Return _Password
    End Get
    Set(ByVal value As String)
      _Password = value
    End Set
  End Property

  'return of set HashAlgorithm for generating
  'Reference and serialkey
  Property HashAlgorithm() As HashAlgorithms
    Get
      Return _HashAlgorithm
    End Get
    Set(ByVal value As HashAlgorithms)
      _HashAlgorithm = value
    End Set
  End Property

  'returns end-user's hard disk serial number
  ReadOnly Property HardDiskSerial() As String
    Get
      Return _DiskSerial
    End Get
  End Property

  'use this property to check if your software is registered
  ReadOnly Property isRegistered() As Boolean

    Get

      If Trim(_RegistryBaseKey) = "" Then _
        Throw New Exception("Specify a BaseRegistryKey")

      Try
        Dim objRegistry As RegistryKey
        Select Case _RegistryHive
          Case RegistryHives.CURRENTUSER
            objRegistry = Registry.CurrentUser
          Case RegistryHives.LOCALMACHINE
            objRegistry = Registry.LocalMachine
          Case Else
            objRegistry = Registry.CurrentUser
        End Select

        Dim RK As RegistryKey = objRegistry. _
          OpenSubKey("Software\" & _RegistryBaseKey)
        Dim SKey As String = RK.GetValue(_RegistrySubKey)
        If SKey = SerialKey Then
          Return True
        Else
          Return False
        End If
      Catch ex As NullReferenceException
        Return False
      End Try

    End Get

  End Property

  'Base Registry key:
  'usually your software name
  Property RegistryBaseKey() As String
    Get
      Return _RegistryBaseKey
    End Get
    Set(ByVal value As String)
      _RegistryBaseKey = value
    End Set
  End Property

  'a sub key of BaseRegistryKey to store
  'the software's serial key
  Property RegistrySubKey() As String
    Get
      Return _RegistrySubKey
    End Get
    Set(ByVal value As String)
      _RegistrySubKey = value
    End Set
  End Property

  Property ReferenceLength() As Integer
    Get
      Return _CustRefLength
    End Get
    Set(ByVal value As Integer)
      If value >= 6 And value <= 20 Then
        _CustRefLength = value
      Else
        Throw New Exception("ReferenceLength cannot be less than 6 or more than 20 characters")
      End If
    End Set
  End Property

  'use this method in your software's registration form's
  'OK button, and pass it the serial key entered by the end-user
  'as argument, returns True for a successful registration, False
  'for unsuccessful
  Function Register(ByVal strSerialKey As String) As Boolean
    GenerateCodes()

    Dim objRegistry As RegistryKey
    Select Case _RegistryHive

      Case RegistryHives.CURRENTUSER
        objRegistry = Registry.CurrentUser
      Case RegistryHives.LOCALMACHINE
        objRegistry = Registry.LocalMachine
      Case Else
        objRegistry = Registry.CurrentUser
    End Select

    If strSerialKey = _SerialKey Then
      Dim RK As RegistryKey = objRegistry. _
       OpenSubKey("Software", True)
      RK.CreateSubKey(_RegistryBaseKey). _
        SetValue(_RegistrySubKey, _SerialKey)
      Return True
    Else
      Return False
    End If
  End Function

  'returns customer reference based on appname,password,harddisk serial
  Overridable Property Reference() As String
    Get
      GenerateCodes()
      Return _CustRef
    End Get
    Set(ByVal value As String)
      If value.Length >= 6 And value.Length <= 20 Then
        _CustRef = value
      Else
        Throw New Exception("Reference cannot be less than 6 or more than 20 characters")
      End If
    End Set
  End Property

  'returns serial key based on Reference
  Overridable ReadOnly Property SerialKey() As String
    Get
      GenerateCodes()
      Return _SerialKey
    End Get
  End Property

  Public Overridable Function CancelRegistration() As Boolean
    'you can use this function for testing purposes
    'should you need to remove any test entries from
    'the windows registry where registration information is stored

    Dim objRegistry As RegistryKey
    Select Case _RegistryHive
      Case RegistryHives.CURRENTUSER
        objRegistry = Registry.CurrentUser
      Case RegistryHives.LOCALMACHINE
        objRegistry = Registry.LocalMachine
      Case Else
        objRegistry = Registry.CurrentUser
    End Select

    Dim RK As RegistryKey = objRegistry.OpenSubKey _
      ("Software", True)
    MsgBox("The followinag key and all its subkeys will be deleted:" & vbCrLf & _
      RK.ToString & "\" & _RegistryBaseKey, MsgBoxStyle.Information)
    RK.DeleteSubKeyTree(_RegistryBaseKey)

  End Function

  Protected Sub GenerateCodes()
    'generate Reference based on 
    'appname,password,and harddisk serial
    'generate serial key based on Reference
    Dim Hash As HashAlgorithm
    Dim HashBytes() As Byte
    Dim UNIEncoding As New UnicodeEncoding
    Dim temp As String

    'load the user selected hash-algorithm
    Select Case _HashAlgorithm
      Case HashAlgorithms.MD5
        Hash = New MD5CryptoServiceProvider
      Case HashAlgorithms.SHA1
        Hash = New SHA1CryptoServiceProvider
      Case HashAlgorithms.SHA256
        Hash = New SHA256Managed
      Case HashAlgorithms.SHA384
        Hash = New SHA384Managed
      Case HashAlgorithms.SHA512
        Hash = New SHA512Managed
      Case Else 'default hash algorithm
        Hash = New MD5CryptoServiceProvider
    End Select

    'generate hash using appname, password, and HD serial no.
    temp = _AppName & _Password & _DiskSerial
    HashBytes = Hash.ComputeHash(UNIEncoding.GetBytes(temp))

    _CustRef = Convert.ToBase64String(HashBytes)
    'if longer than 20 chars, trim it to 20 chars
    If _CustRef.Trim.Length > _CustRefLength Then _CustRef = Left(_CustRef, _CustRefLength)

    'convert it to upper case
    _CustRef = UCase(_CustRef)

    'clear the hash array
    Array.Clear(HashBytes, 0, HashBytes.Length)

    'use the above generated Reference to generate a 20 characters 
    'serial key for the end user
    temp = _CustRef & _AppName & _Password
    HashBytes = Hash.ComputeHash(UNIEncoding.GetBytes(temp))

    _SerialKey = Convert.ToBase64String(HashBytes)
    If _SerialKey.Trim.Length > 20 Then _SerialKey = Left(_SerialKey, 20)
    _SerialKey = UCase(_SerialKey)

  End Sub


End Class
